grr/server/grr_response_server/bin/grrafana.py

```diff
@@ -51,22 +73,134 @@ class JSONResponse(werkzeug_wrappers_json.JSONMixin,
 
   def __init__(self, response=None, *args, **kwargs) -> None:
     kwargs["mimetype"] = JSON_MIME_TYPE
-    if response is not None and not isinstance(response, werkzeug_wsgi.ClosingIterator):
+    if response is not None and not isinstance(response,
+                                               werkzeug_wsgi.ClosingIterator):
       response = json.dumps(response)
     super().__init__(response, *args, **kwargs)
 
 
+class Metric(abc.ABC):
+  """A single metric that can be fetched from a
+  Fleetspeak-based GRR deployment."""
+
+  @abc.abstractmethod
+  def __init__(self, name: str) -> None:
+    """Initializes a new metric."""
+    self._name = name
+
+  def GetName(self) -> str:
+    """Returns the name of the metric."""
+    return self._name
+
+  @abc.abstractmethod
+  def ProcessQuery(self, req: JSONRequest):
+    """Processes the request JSON data and returns
+    data that can be read by Grafana (by JSON Datasource plugin)."""
+    pass
+
+
+class ClientResourceUsageMetric(Metric):
+  """A metric that reprsents resource usage data for a single client,
+  fetched from Fleetspeak database."""
+
+  def __init__(self,
+               name: str,
+               record_values_extract_fn: Callable[[List[resource_pb2.ClientResourceUsageRecord]], List[float]]) -> None:
+    super().__init__(name)
+    self._record_values_extract_fn = record_values_extract_fn
+
+  def ProcessQuery(self, req: JSONRequest) -> _TargetWithDatapoints:
+    client_id = req["scopedVars"]["ClientID"]["value"]
+    start_range_ts = timeToProtoTimestamp(req["range"]["from"])
+    end_range_ts = timeToProtoTimestamp(req["range"]["to"])
+    records_list = fleetspeak_utils.FetchClientResourceUsageRecords(
+        client_id, start_range_ts, end_range_ts)
+    record_values = self._record_values_extract_fn(records_list)
+    datapoints = [_Datapoint(nanos=v,
+                             value=r.server_timestamp.seconds * 1000 +
+                                   r.server_timestamp.nanos // 1000000)
+                  for (v, r) in zip(record_values, records_list)]
+    return _TargetWithDatapoints(target=self._name, datapoints=datapoints)
+
+
+class ClientsStatisticsMetric(Metric):
+
+  def __init__(self,
+               name: str,
+               get_fleet_stats_fn: Callable[[frozenset], fleet_utils.FleetStats],
```

Comment:
Prefer not to use raw `frozenset` in type annotations, but rather generic [`FrozenSet`][1] with specific type.

[1]: https://docs.python.org/3/library/typing.html#typing.FrozenSet